### DISCUSSION TITLE
Reference for how `abs`, `maximum`, and others forward diff. smoothing techniques

### DISCUSSION THREAD

**Discussion by jeertmans** [jax-ml/jax#13350] [Q&A] [answered]
Hello everyone,

Functions with non-smooth outputs, such as `abs(x)` when `x = 0`, or `maximum(x, y)` when `x = y` are *smoothed* at those critical points. Indeed, they should have undefined derivative, but Jax returns a defined value.

For example, `maximum(x, y)` returns `(x.dx + y.dx) / 2` when `x = y`. On the other hand, `abs(x)` totally ignores the case when `x = 0` (which should be treated differently).

I understand that working with NaNs is not convenient, but I wanted to know it there was a reference paper (or else) that motivates their choices regarding those changes?

**Comment by froystig** [comment, marked as answer]
You're right that these functions (`abs`, `maximum`) aren't differentiable at these points. They do, however, have [subderivatives](https://en.wikipedia.org/wiki/Subderivative) at those points, and jax answers with that, at least.

Why return anything at all, rather than error? One answer is that often this is what users expect. Try another autodiff library and you'll see similar output.

Why return subderivatives when possible? One answer is that autodiff is often used for optimization, and [subgradient methods](https://en.wikipedia.org/wiki/Subgradient_method) minimize convex functions under standard conditions. People often use convex optimization methods outside of these settings, such as in non-smooth convex optimization (e.g. training a linear predictor under the [hinge loss](https://en.wikipedia.org/wiki/Hinge_loss)), or in training deep nets (where objectives are neither convex nor differentiable everywhere). This is another way of saying, again, that often this is what users expect.

Within jax, we could imagine implementing a "strict AD" that raises errors when a primitive encounters a non-differentiable point. Roughly speaking, we could err early in JVPs, and rely on our [decomposition of AD](https://arxiv.org/abs/2105.09469) to handle the rest. But as far as I know, we've never encountered a real request for this.

It's also tricky to do this in a way that isn't prohibitively strict. AD breaks derivatives of large expressions down to derivatives of basic primitives, strung back together by composition rules. Sometimes primitive subexpressions are not differentiable whereas their compound expression is. For example, using `maximum`, the following is an unfortunate way to express the identity function (which is differentiable everywhere!):

```python
>>> f = lambda x: jnp.maximum(0, x) - jnp.maximum(0, -x)
>>> print(jax.grad(f)(1.))
1.0
>>> print(jax.grad(f)(-1.))
1.0
>>> print(jax.grad(f)(0.))
1.0
```

A simple "strict" AD implementation might err at the origin because either subexpression involving `maximum` isn't strictly differentiable. But here we give the answer you'd arguably like to see overall.

**Reply by jeertmans** [reply]
Many thanks for the comprehensive answer @froystig!
